cylc/flow/cycling/__init__.py

```diff
@@ -236,15 +261,39 @@ def __add__(self, other):
             raise CyclerTypeError(self.TYPE, self, other.TYPE, other)
         return self.add(other)
 
-    def __cmp__(self, other):
+    def __cmp__(self, other: 'IntervalBase') -> int:
         # Compare self to other (interval).
+        # TODO: refactor as __cmp__ not used in python3
         if other is None:
             return -1
+        if not isinstance(other, IntervalBase):
+            return NotImplemented
+        # Note: apparently we can compare different subclasses of IntervalBase:
         if self.TYPE != other.TYPE:
             return cmp(self.TYPE_SORT_KEY, other.TYPE_SORT_KEY)
         if self.value == other.value:
             return 0
-        return self.cmp_(other)
+        return self._cmp(other)
+
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, self.__class__):
+            return self.__cmp__(other) == 0
+        return NotImplemented
+
+    def __lt__(self, other: 'IntervalBase') -> bool:
+        return self.__cmp__(other) == -1
+
+    def __le__(self, other: 'IntervalBase') -> bool:
+        return self.__cmp__(other) <= 0
+
+    def __gt__(self, other: 'IntervalBase') -> bool:
+        return self.__cmp__(other) == 1
+
+    def __ge__(self, other: 'IntervalBase') -> bool:
+        return self.__cmp__(other) >= 0
+
+    def __hash__(self) -> int:
+        return hash(self.value)
```

Comment:
I don't think `__hash__` was being implemented in `IntegerInterval` or `ISO8601Interval` before this, so two ostensibly equal intervals would have had different hashes (per-instance hash)